engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/job/EngineJobExecutionActor.scala

```diff
@@ -140,29 +141,41 @@ class EngineJobExecutionActor(replyTo: ActorRef,
   when(Pending) {
     case Event(Execute, NoData) =>
       increment(NonEmptyList("jobs", List("ejea", "executing", "starting")))
-      requestExecutionToken()
-      goto(RequestingExecutionToken)
-  }
-
-  when(RequestingExecutionToken) {
-    case Event(JobExecutionTokenDispensed, NoData) =>
-      replyTo ! JobStarting(jobDescriptorKey)
       if (restarting) {
-        val jobStoreKey = jobDescriptorKey.toJobStoreKey(workflowIdForLogging)
-        jobStoreActor ! QueryJobCompletion(jobStoreKey, jobDescriptorKey.call.outputPorts.toSeq)
-        goto(CheckingJobStore)
+        requestRestartCheckToken()
+        goto(RequestingRestartCheckToken)
       } else {
-        requestValueStore()
+        requestExecutionToken()
+        goto(RequestingExecutionToken)
       }
   }
 
+  // This condition only applies for restarts
+  when(RequestingRestartCheckToken) {
+    case Event(JobTokenDispensed, NoData) =>
+      replyTo ! JobStarting(jobDescriptorKey)
+      val jobStoreKey = jobDescriptorKey.toJobStoreKey(workflowIdForLogging)
+      jobStoreActor ! QueryJobCompletion(jobStoreKey, jobDescriptorKey.call.outputPorts.toSeq)
+      goto(CheckingJobStore)
+  }
+
+  when(RequestingExecutionToken) {
+    case Event(JobTokenDispensed, NoData) =>
+      if (!restarting)
```

Comment:
Is this check redundant if we no longer request execution tokens on restarts?

Reply:
If the restart check finds out that the job in fact never started then it asks for an execution token and waits here for it to arrive.

Reply:
Ah so the logic is always try to restart, and if can't restart, start?

Reply:
yes that's right.